compiler/vm/vmdef.nim

```diff
@@ -213,3 +561,175 @@ template regC*(x: TInstr): TRegister = TRegister(x.TInstrType shr regCShift and
 template regBx*(x: TInstr): int = (x.TInstrType shr regBxShift and regBxMask).int
 
 template jmpDiff*(x: TInstr): int = regBx(x) - wordExcess
+
+func makeHandle*(rh: ptr Atom, sizeInBytes: uint): MemoryHandle =
+  MemoryHandle(p: rh, l: sizeInBytes.int)
+
+func makeTHandle*(h: MemoryHandle, typ: PVmType): TypedHandle {.inline.} =
+  assert typ.sizeInBytes <= uint(h.l)
+  (MemoryHandle(p: h.p, l: int(typ.sizeInBytes)), typ)
+
+
+func makeTHandle*(h: ptr Atom, typ: PVmType): TypedHandle {.inline.} =
+  (MemoryHandle(p: h, l: int(typ.sizeInBytes)), typ)
+
+func toPointer*(h: MemoryHandle): ptr Atom =
+  h.p
+
+template isNil*(h: MemoryHandle): bool = h.p == nil
+
+template invalidHandle*: untyped = MemoryHandle(p: nil, l: 0)
+template invalidTypedHandle*: untyped = (h: invalidHandle, typ: nil)
+
+template isValid*(t: PVmType): bool = t != nil
+
+func getRawPointer*(h: MemoryHandle): pointer {.inline.} =
+  ## Don't use this directly, it's only here for ``vmmemory``
+  h.p
+
+template len*(h: MemoryHandle): untyped = h.l
+
+# Safe accessors for `VmObjecAtom`'s fields
+template accessor(n, k): untyped {.dirty.} =
+  template `n`*(a: VmObjectAtom): untyped = 
+    assert a.kind == k
+    cast[ptr Atom](unsafeAddr a.atom).n
+
+accessor(intVal, akInt)
+accessor(floatVal, akFloat)
+accessor(ptrVal, akPtr)
+accessor(setVal, akSet)
+accessor(strVal, akString)
+accessor(seqVal, akSeq)
+accessor(refVal, akRef)
+accessor(callableVal, akCallable)
+accessor(nodeVal, akPNode)
+
+
+func newVmString*(data: MemoryHandle, len: Natural): VmString {.inline.} =
+  VmString(VmSeq(data: data, length: len))
+
+func len*(s: VmString): int {.inline.} = VmSeq(s).length
+
+func data*(s: VmString): MemoryHandle {.inline.} = VmSeq(s).data
+
+# Prevent copying of `TFullReg` at compile so that it's
+# invariants don't get violated by an accidental copy
+proc `=`(a: var TFullReg, b: TFullReg) {.error.}
+
+func getHandleToValue*(a: VmObjectAtom): TypedHandle =
+  var h = 
+    case a.kind
+    of realAtomKinds:
+      MemoryHandle(p: cast[ptr Atom](unsafeAddr a.atom), l: sizeof(Atom))
+    else:
+      a.obj
+
+  makeTHandle(h, a.typ)
+
+template `atomVal`*(r: TFullReg): untyped =
+  r.handle.h.p[]
+
+
+template isValid*(h: MemoryHandle, len: uint): bool = h.p != nil and int(len) <=% h.l 
+template isValid*(handle: TypedHandle): bool = 
+  let x = handle
+  isValid(x.h, x.typ.sizeInBytes)
```

Comment:
That's a lot less checking than I'd feel comfortable with, especially for a function called `isValid` that guards all `deref` in `vmmemory` :P.

If it's just a kind of sanity check for `vmmemory`, I'd recommend moving it there instead and maybe call it `isNotNil` :D

Reply:
Yea, it's just a sanity check and debugging aid. Moving it to `vmmemory` would require exporting the fields of `MemoryHandle`, which is why I didn't do it.

The vm modules need to be cleaned up and reorganized in general

Reply:
I'd leave a to-do in there, but a smaller public API is preferable than premature organization.